databricks/koalas/groupby.py

```diff
@@ -640,6 +640,65 @@ def diff(self, periods=1):
             lambda sg: sg._kser._diff(periods, part_cols=sg._groupkeys_scols)
         )
 
+    def cumcount(self, ascending=True):
+        """
+        Number each item in each group from 0 to the length of that group - 1.
+
+        Essentially this is equivalent to
+
+        >>> self.apply(lambda x: pd.Series(np.arange(len(x)), x.index)) # doctest: +SKIP
+
+        Parameters
+        ----------
+        ascending : bool, default True
+            If False, number in reverse, from length of group - 1 to 0.
+
+        Returns
+        -------
+        Series
+            Sequence number of each element within each group.
+
+        Examples
+        --------
+
+        >>> df = ks.DataFrame([['a'], ['a'], ['a'], ['b'], ['b'], ['a']],
+        ...                   columns=['A'])
+        >>> df
+           A
+        0  a
+        1  a
+        2  a
+        3  b
+        4  b
+        5  a
+        >>> df.groupby('A').cumcount().sort_index()
+        0    0
+        1    1
+        2    2
+        3    0
+        4    1
+        5    3
+        Name: 0, dtype: int64
+        >>> df.groupby('A').cumcount(ascending=False).sort_index()
+        0    3
+        1    2
+        2    1
+        3    1
+        4    0
+        5    0
+        Name: 0, dtype: int64
+
+        """
+        ret = self._apply_series_op(
+            lambda sg: sg._kser._cum(
+                F.count, True, part_cols=sg._groupkeys_scols, ascending=ascending
+            ),
+            should_resolve=True,
+        )
+        ret -= 1
+        # Cast columns to ``"int64"`` to match `pandas.core.groupby.GroupBy.cumcount`.
+        return ret.max(axis=1).astype('int64')
```

Comment:
I don't think using `_apply_series_op` is a good idea here.

How about:

```diff
@@ -689,15 +689,9 @@ class GroupBy(object, metaclass=ABCMeta):
         Name: 0, dtype: int64

         """
-        ret = self._apply_series_op(
-            lambda sg: sg._kser._cum(
-                F.count, True, part_cols=sg._groupkeys_scols, ascending=ascending
-            ),
-            should_resolve=True,
-        )
-        ret -= 1
-        # Cast columns to ``"int64"`` to match `pandas.core.groupby.GroupBy.cumcount`.
-        return ret.max(axis=1).astype('int64')
+        ret = self._groupkeys[0]._cum(F.count, True, self._groupkeys_scols, ascending=ascending) - 1
+        internal = ret._internal.resolved_copy
+        return first_series(DataFrame(internal))

     def cummax(self):
         """
```

still need to check the behavior when `self._groupkeys[0]` contains null values.

Reply:
This did indeed not work for null values, but I found another possibility with `F.row_count` instead. It even works with the empty array example and matches the pandas implementation now!
By default the returning values seem to be `int32` now. Is that fine as well or would be the cast to `int64` be better to match pandas?

Reply:
What happened when null values with `F.count`?

Reply:
Null values are explicitely set [in series._cum](https://github.com/databricks/koalas/blob/master/databricks/koalas/series.py#L5057), when avoiding that, and just executing the `.otherwise` part there, the null values are -1 instead of counting upwards. It might work to fill the nulls with a non-existing value in that column and then using count .. this seems to work fine so far:
```
keys = self._groupkeys[0].copy()
keys.fillna(keys.max() + 1, inplace=True)
ret = (
            keys._cum(
                F.count,
                True,
                nocolarg=False,
                part_cols=self._groupkeys_scols,
                ascending=ascending,
            )
            - 1
        )
internal = ret._internal.resolved_copy
return first_series(DataFrame(internal))
```
Would that be an alternative?

Reply:
`keys.max()` yields one extra Spark job, also `fillna` needs to change the value according to the data type.
How about:

```py
ret = (
            self._groupkeys[0].spark.transform(lambda _: F.lit(0))._cum(
                F.count,
                True,
                part_cols=self._groupkeys_scols,
                ascending=ascending,
            )
            - 1
        )
```

Btw, we won't need `nocolarg` parameter anymore. Could you revert the change?